src/output-json.c

```diff
@@ -707,6 +710,76 @@ void CreateJSONFlowId(json_t *js, const Flow *f)
     }
 }
 
+static inline void JSONFormatAndAddMACAddr(json_t *parent, const char *key,
+                                   uint8_t *val, bool is_array)
+{
+    char eth_addr[19];
+    (void) snprintf(eth_addr, 19, "%02x:%02x:%02x:%02x:%02x:%02x",
+                    val[0], val[1], val[2], val[3], val[4], val[5]);
+    if (is_array) {
+        json_array_append_new(parent, json_string(eth_addr));
+    } else {
+        json_object_set_new(parent, key, json_string(eth_addr));
+    }
+}
+
+/* only required to traverse the MAC address set */
+typedef struct JSONMACAddrInfo {
+    json_t *src, *dst;
+} JSONMACAddrInfo;
+
+static int JSONFlowAppendMACAddrs(uint8_t *val, MacSetSide side, void *data)
+{
+    JSONMACAddrInfo *info = (JSONMACAddrInfo*) data;
+    if (side == MAC_SET_DST) {
+        JSONFormatAndAddMACAddr(info->dst, NULL, val, true);
+    } else {
+        JSONFormatAndAddMACAddr(info->src, NULL, val, true);
+    }
+    return 0;
+}
+
+int CreateJSONEther(json_t *parent, const Packet *p, const Flow *f)
+{
+    json_t *ejs = json_object();
+    if (unlikely(ejs == NULL))
+        return 0;
+    if (p == NULL) {
+        /* we are creating an ether object in a flow context, so we need to
+           append to arrays */
+        if (f != NULL && f->macset && MacSetSize(f->macset) > 0) {
+            JSONMACAddrInfo info;
+            info.dst = json_array();
+            info.src = json_array();
+            int ret = MacSetForEach(f->macset, JSONFlowAppendMACAddrs, &info);
+            if (unlikely(ret != 0)) {
+                /* should not happen, JSONFlowAppendMACAddrs is sane */
+                json_decref(info.dst);
+                json_decref(info.src);
+                json_decref(ejs);
+                return ret;
+            }
+            json_object_set_new(ejs, "dest_macs", info.dst);
+            json_object_set_new(ejs, "src_macs", info.src);
+        }
+    } else {
+        /* this is a packet context, so we need to add scalar fields */
+        uint8_t *src, *dst;
+        if ((PKT_IS_TOCLIENT(p))) {
+            src = p->ethh->eth_dst;
+            dst = p->ethh->eth_src;
+        } else {
+            src = p->ethh->eth_src;
+            dst = p->ethh->eth_dst;
+        }
+        json_object_set_new(ejs, "type", json_integer(ntohs(p->ethh->eth_type)));
+        JSONFormatAndAddMACAddr(ejs, "src_mac", src, false);
```

Comment:
Would it be possible to use `JSONFlowAppendMACAddrs` instead of calling `JSONFormatAndAddMACAddr`? Perhaps rename `JSONFlowAppend` --> `JSONMACAppend`?

This would reduce reuse code and reduce the # of calls to `JSONFormatAndAddMACAddr`.

Reply:
Hm. I could but to me that would feel unnatural, TBH. `JSONFlowAppendMACAddrs` is just meant to be a one-off utility function to match the `MacSetIteratorFunc` and IMHO only makes sense where `MacSetForEach` is called. 
`JSONFormatAndAddMACAddr` is what I feel one would want to use in general, and also it is inlined, saving us one real function call in the packet context.

I can rename `JSONFlowAppendMACAddrs` to `MacSetIterateToJSON` to make it clearer and reduce confusion, what do you think?

Reply:
Leave things as they are. Just wanted to check. Since you're inlining the function, it should be ok.